Adding tutorial on monitoring Geth (#2241)
ethereum/ethereum-org-website: src/content/developers/tutorials/monitoring-geth-with-influxdb-and-grafana/index.md

---
title: Monitoring Geth with InfluxDB and Grafana
description: 
author: "Mario Havel"
tags: ["clients", "geth", "nodes"]
skill: intermediate
lang: en
sidebar: true
---

> Monitoring your node is a best practice which can help you to be informed about its performance, status, identify potential issues and even alert you if some problem occurs. Client implementations including [Geth](https://geth.ethereum.org/) offer thorough metrics collection and reporting. 

Reviewer: This tutorial will help you set up monitoring for your Geth node so you can better understand its performance and identify potential problems.

## Prerequisites {#prerequisites}

> * Tutorial assumes that you are already running an instance of Geth client. 

Reviewer: - You should already be running an instance of Geth.

* Most of the steps and examples are for linux environment, basic terminal knowledge will be helpful. 

> * To get familiar with Geth's suite of metrics, its use for debugging and why it might be important to you, check talk [Monitoring an Ethereum infrastructure by Péter Szilágyi](https://www.youtube.com/watch?v=cOBab8IJMYI).

Reviewer: - Check out this video overview of Geth's suite of metrics: [Monitoring Ethereum infrastructure by Péter Szilágyi](https://www.youtube.com/watch?v=cOBab8IJMYI)

## Monitoring stack {#monitoring-stack}

> Ethereum client collects various metrics from which a chronological database can be created. This database will be then read by data visualizing software. This means we need a stack of different tools working together. There are multiple variations of monitoring stack available with software like Prometheus (pull model), InfluxDB (push model), Telegraf, Grafana, Datadog, Chronograf, etc. 

Reviewer: An Ethereum client collects lots of data which can be read in the form of a chronological database. To make monitoring easier, you can feed this into data visualisation software. There are multiple options available:
- Prometheus (pull model)
- InfluxDB (push model)
- Telegraf
- Grafana
- Datadog
- Chronograf

Can we make these links?

> Preconfigured options worth mentioning are [Geth Prometheus Exporter](https://github.com/hunterlong/gethexporter) which enables easy setup using docker and [Ethbian](https://ethbian.org/index.html) OS for RPi 4 which comes with InfluxDB and Grafana preconfigured. 

Reviewer: There's also [Geth Prometheus Exporter](https://github.com/hunterlong/gethexporter), an option preconfigured with InfluxDB and Grafana. You can set it up easily using docker and [Ethbian](https://ethbian.org/index.html) OS for RPi 4.

> In this tutorial, we will manually setup InfluxDB with Grafana. Expected outcome is Geth pushing metrics data to InfluxDB which creates time series database and Grafana dashboard using this database as a source for graph visualization. Manual setup should teach you to understand the process, alter it and deploy on different environments. 

Reviewer: In this tutorial, we'll set up your Geth client to push data to InfluxDB to create a database and Grafana to create a graph visualisation of the data. Doing it manually will help you understand the process better, alter it, and deploy in different environments.

## Setting up InfluxDB {#setting-up-influxdb}

> First, let's download and install InfluxDB. Various download options can be found at [Influxdata release page](https://portal.influxdata.com/downloads/), pick the one suiting your environment. 

Reviewer: First, let's download and install InfluxDB. Various download options can be found at [Influxdata release page](https://portal.influxdata.com/downloads/). Pick the one that suits your environment.

You can also install it from a [repository](https://repos.influxdata.com/). For example in Debian based distribution: 
```
curl -tlsv1.3 --proto =https -sL https://repos.influxdata.com/influxdb.key | sudo apt-key add 
source /etc/lsb-release
echo "deb https://repos.influxdata.com/${DISTRIB_ID,,} ${DISTRIB_CODENAME} stable" | sudo tee /etc/apt/sources.list.d/influxdb.list
sudo apt update
sudo apt install influxdb -y
sudo systemctl enable influxdb
sudo systemctl start influxdb
sudo apt install influxdb-client
```

> After successfully installing InfluxDB, make sure it's running on background. By default, it is reachable at `localhost:8086`. 

Reviewer: After successfully installing InfluxDB, make sure it's running in the background.

Author: I would argue to leave it there, it is general information and will be used multiple times later in tutorial

> Before using `influx` client, you have to create new user with admin privileges. This user will serve for high level management, creating databases and users. 

Reviewer: Before using `influx` client, you have to create a new user with admin privileges. This user will be able to create databases and users.

```
curl -XPOST "http://localhost:8086/query" --data-urlencode "q=CREATE USER username WITH PASSWORD 'password' WITH ALL PRIVILEGES"
```

> Now you can use influx client to enter [InfluxDB shell](https://docs.influxdata.com/influxdb/v1.8/tools/shell/) with this user. 

Reviewer: "the influx client" or "your influx client"

"the InfluxDB shell" ???

```
influx -username 'username' -password 'password'
```

> Directly communicating with InfluxDB in its shell, you can create database and user for geth metrics. 

Reviewer: By directly communicating with InfluxDB in its shell, you can create databases and users for geth metrics.

```
create database geth
create user geth with password choosepassword
```

Verify created entries with:
```
show databases
show users
```
Leave InfluxDB shell.
```
exit
```

> InfluxDB is running and configured to store metrics from Geth. 

Reviewer: InfluxDB is now running and configured to store metrics from Geth!

## Preparing Geth {#preparing-geth}

> After setting up database, we need to enable metrics collection in Geth. Pay attention to `METRICS AND STATS OPTIONS` in `geth --help`. Multiple options can be found there, in this case we want Geth to push data into InfluxDB. 

Reviewer: After setting up the database, you need to enable metrics collection in Geth. Pay attention to `METRICS AND STATS OPTIONS` in `geth --help`. Multiple options can be found there but in this case we want Geth to push data into InfluxDB.

> Basic setup specifies endpoint where InfluxDB is reachable and authentication for the database.

Reviewer: Can we turn this sentence into an instruction?

```
geth --metrics --metrics.influxdb --metrics.influxdb.endpoint "http://0.0.0.0:8086" --metrics.influxdb.username "geth" --metrics.influxdb.password "chosenpassword"
```

> This flags can be appended to a command starting the client or saved to the configuration file.

Reviewer: This flag? singular?

> You can verify that Geth is successfully pushing data, for instance by listing metrics in database. In InfluxDB shell:

Reviewer: the database?

Author: Given command just lists names of items so I would rather specify that it is just one example 
"Verify that Geth is successfully pushing data, for example by listing metrics using InfluxDB shell:"

```
use geth
show measurements
```
## Setting up Grafana {#setting-up-grafana}

> Next step is installing Grafana which will interpret data graphically. Follow installation process for your environment in Grafana documentation. Make sure to install OSS version if you don't want otherwise. 

Reviewer: the installation process

> Next step is installing Grafana which will interpret data graphically. Follow installation process for your environment in Grafana documentation. Make sure to install OSS version if you don't want otherwise. 

Reviewer: Don't understand the last sentence.

Author: Forgot to add link to docs. Grafana offers OSS and Enterprise version, documentation might mislead to installing enterprise https://grafana.com/docs/grafana/latest/installation/

> Example installation steps for Debian distributions using repository: 

Reviewer: Can we make this an instruction?

```
curl -tlsv1.3 --proto =https -sL https://packages.grafana.com/gpg.key | sudo apt-key add - 
echo "deb https://packages.grafana.com/oss/deb stable main" | sudo tee -a /etc/apt/sources.list.d/grafana.list
sudo apt update
sudo apt install grafana
sudo systemctl enable grafana-server
sudo systemctl start grafana-server
```

> Successfully installed Grafana is running, by default is reachable at `localhost:3000`.

Reviewer: When you've got Grafana running, it should be reachable at `localhost:3000`.

> Use preferred browser to access it and login with default credentials user ‘admin’ and password ‘admin’. You will be prompted to change the default password, type your new password and save. 

Reviewer: "your preferred browser"
"your default credentials"

"When prompted, change the default password and save. "

![](./grafana1.png)

> You will be redirected to default Grafana home page. First thing to setup in Grafana is source of data. Click on on configuration icon on left bar and select Data sources. 

Reviewer: "the Grafana homepage"

First, set up your source data. Click on the configuration icon in the left bar and select "Data sources".

![](./grafana2.png)

> There aren't any data sources created yet, click on Add data source to define one. 

Reviewer: "Add data source"

![](./grafana3.png)

> For this setup, select InfluxDB option and proceed. 

Reviewer: For this setup, select "InfluxDB" and proceed.

![](./grafana4.png)

> Data source configuration is pretty straight forward if you are running tools on the same machine. You need to set address of InfluxDB and details for accessing database. Refer to the picture below.  

Reviewer: "the InfluxDB address" 
"the database"

![](./grafana5.png)

> If everything is filled correctly and InfluxDB is reachable, click on Save & Test and confirmation that source is working will pop up. 

Reviewer: If everything is complete and InfluxDB is reachable, click on "Save and test" and wait for the confirmation to pop up.

![](./grafana6.png)

> Grafana is set up to read data from InfluxDB. Now you need to create dashboard which will interpret and display it. Dashboards properties are encoded in JSON files which can be created by anybody and easily imported. On left bar, click on Create and Import. 

Reviewer: Grafana is now set up to read data from InfluxDB. Now you need to create a dashboard which will interpret and display it. Dashboards properties are encoded in JSON files which can be created by anybody and easily imported. On the left bar, click on "Create and Import".

![](./grafana7.png)

!!Add dashboard to grafana website for easier import? 

> Dashboard for monitoring Geth client is provided, just copy [this JSON file](./gethdashboard.json) and insert it to Import page in Grafana. After successfully saving the dashboard, it should look like this. 

Reviewer: For a Geth monitoring dashboard, just copy [this JSON file](./gethdashboard.json) and insert it in the "Import page" in Grafana. After saving the dashboard, it should look like this:

Reviewer 2: Just a note, this link didn't work correctly for me in the netlify preview, getting parsed to `/en/{hash}/gethdashboard.json`... Can take a look at this when the rest is finalized.

![](./grafana8.png)

> Dashboards can be interactively modified. Each panel can be edited, moved, removed or added, configuration can be changed, everything is opened and up to user. To learn more about how dashboards work, refer to [Grafana documentation](https://grafana.com/docs/grafana/latest/dashboards/). 

Reviewer: You can modify your dashboards. Each panel can be edited, moved, removed or added. You can change your configurations. It's up to you! To learn more about how dashboards work, refer to [Grafana's documentation](https://grafana.com/docs/grafana/latest/dashboards/).

> Another feature you might be interested in is [Alerting](https://grafana.com/docs/grafana/latest/alerting/). It enables you to trigger alert notification when metric reach specific value, various communication channels are supported. 

Reviewer: You might also be interested in [Alerting](https://grafana.com/docs/grafana/latest/alerting/). This lets you set up alert notifications for when metrics reach certain values. Various communication channels are supported.

Example installation steps for Debian distributions using repository: 
```
curl -tlsv1.3 --proto =https -sL https://packages.grafana.com/gpg.key | sudo apt-key add - 
echo "deb https://packages.grafana.com/oss/deb stable main" | sudo tee -a /etc/apt/sources.list.d/grafana.list
sudo apt update
sudo apt install grafana
sudo systemctl enable grafana-server
sudo systemctl start grafana-server
```
Successfully installed Grafana is running, by default is reachable at `localhost:3000`.
Use preferred browser to access it and login with default credentials user ‘admin’ and password ‘admin’. You will be prompted to change the default password, type your new password and save. 

![](./grafana1.png)

You will be redirected to default Grafana home page. First thing to setup in Grafana is source of data. Click on on configuration icon on left bar and select Data sources. 

![](./grafana2.png)

There aren't any data sources created yet, click on Add data source to define one. 

![](./grafana3.png)

For this setup, select InfluxDB option and proceed. 

![](./grafana4.png)

> Data source configuration is pretty straight forward if you are running tools on the same machine. You need to set address of InfluxDB and details for accessing database. Refer to the picture below.  

Reviewer 3: ```suggestion
Data source configuration is pretty straight forward if you are running tools on the same machine. You need to set the InfluxDB address and details for accessing the database. Refer to the picture below.  
```

![](./grafana5.png)

> If everything is filled correctly and InfluxDB is reachable, click on Save & Test and confirmation that source is working will pop up. 

Reviewer 3: ```suggestion
If everything is complete and InfluxDB is reachable, click on "Save and test" and wait for the confirmation to pop up.
```

![](./grafana6.png)

> Grafana is set up to read data from InfluxDB. Now you need to create dashboard which will interpret and display it. Dashboards properties are encoded in JSON files which can be created by anybody and easily imported. On left bar, click on Create and Import. 

Reviewer 3: ```suggestion
Grafana is now set up to read data from InfluxDB. Now you need to create a dashboard which will interpret and display it. Dashboards properties are encoded in JSON files which can be created by anybody and easily imported. On the left bar, click on "Create and Import".
```

![](./grafana7.png)

!!Add dashboard to grafana website for easier import? 

> Dashboard for monitoring Geth client is provided, just copy [this JSON file](./gethdashboard.json) and insert it to Import page in Grafana. After successfully saving the dashboard, it should look like this. 

Reviewer 3: ```suggestion
For a Geth monitoring dashboard, just copy [this JSON file](./gethdashboard.json) and insert it in the "Import page" in Grafana. After saving the dashboard, it should look like this:```

![](./grafana8.png)

> Dashboards can be interactively modified. Each panel can be edited, moved, removed or added, configuration can be changed, everything is opened and up to user. To learn more about how dashboards work, refer to [Grafana documentation](https://grafana.com/docs/grafana/latest/dashboards/). 

Reviewer 3: ```suggestion
ou can modify your dashboards. Each panel can be edited, moved, removed or added. You can change your configurations. It's up to you! To learn more about how dashboards work, refer to [Grafana's documentation](https://grafana.com/docs/grafana/latest/dashboards/). 
```

> Another feature you might be interested in is [Alerting](https://grafana.com/docs/grafana/latest/alerting/). It enables you to trigger alert notification when metric reach specific value, various communication channels are supported. 

Reviewer 3: ```suggestion
You might also be interested in [Alerting](https://grafana.com/docs/grafana/latest/alerting/). This lets you set up alert notifications for when metrics reach certain values. Various communication channels are supported.
```
